src/lib/Imp.hs

```diff
@@ -143,73 +177,161 @@ toImpHof env (maybeDest, hof) = do
+      extend ((mempty, x), (y, z))
+      return a
 
-type Dest = Atom
+-- === Destination type ===
+--
+-- Desinations are an analog to regular Atoms, except they guarantee that the "leaf"
+-- elements are scalar tables with non-aliasing memory. The Dest data structure is a
+-- view that reverses the structure-of-arrays transformation we apply. When we go to
+-- the atom world, we use table lambdas for the same purpose, but destinations are
+-- a little easier to work with, since they can't contain arbitrary code.
+
+-- This takes a type parameter only so that we can derive Traversalbe. Use Dest instead.
+data DestP ref = DFor   (Abs (DestP ref))
+               | DPair  (DestP ref) (DestP ref)
+               | DUnit
+               | DRef   ref
+               | DAsIdx Type (DestP ref)
+                 deriving (Show, Functor, Foldable, Traversable)
+type Dest = DestP IVar
 type WithDest a = (Maybe Dest, a)
 
+instance HasVars Dest where
+  freeVars dest = case dest of
+    DFor abs               -> freeVars abs
+    DPair l r              -> freeVars l <> freeVars r
+    DUnit                  -> mempty
+    DRef (_ :> IValType _) -> mempty
+    DRef (_ :> IRefType t) -> freeVars t
+    DAsIdx t d             -> freeVars t <> freeVars d
+
+  subst env dest = case dest of
+    DFor abs               -> DFor (subst env abs)
+    DPair l r              -> DPair (subst env l) (subst env r)
+    DUnit                  -> dest
+    DRef (_ :> IValType _) -> dest
+    DRef (v :> IRefType t) -> DRef (v :> IRefType (subst env t))
+    DAsIdx t d             -> DAsIdx (subst env t) (subst env d)
+
+destGet :: Dest -> IExpr -> ImpM Dest
+destGet dest i = case dest of
+  DFor a@(Abs v _) -> do
+    idx <- intToIndex (varType v) i
+    traverse (\ref -> toIVar <$> (IVar ref) `impGet` i) $ applyAbs a idx
+  _ -> error $ "Indexing into a non-array dest"
+  where toIVar ~(IVar v) = v
+
+-- XXX: This should only be called when it is known that the dest will _never_
+--      be modified again, because it doesn't copy the state!
+destToAtom :: Dest -> ImpM Atom
+destToAtom dest = destToAtomScalarAction loadScalarRef dest
+  where
+    loadScalarRef ref = toScalarAtom (BaseTy b) <$> (load $ IVar ref)
+      where ~(_ :> IRefType (BaseTy b)) = ref
+
+destToAtomScalarAction :: (IVar -> ImpM Atom) -> Dest -> ImpM Atom
+destToAtomScalarAction fScalar dest = do
+  scope <- looks $ fst . snd
+  -- FIXME: This does not propagate the scope back!
+  (atom, decls) <- runEmbedT (destToAtom' fScalar [] dest) scope
+  case decls of
+    [] -> return $ atom
+    _  -> error "Unexpected decls"
+
+destToAtom' :: (IVar -> ImpM Atom) -> [Var] -> Dest -> EmbedT ImpM Atom
+destToAtom' fScalar forVars dest = case dest of
+  DFor (Abs v d) -> do
+    scope <- looks $ fst . snd
+    v' <- lift $ case v of
+      (NoName :> t) -> freshVar ("idx" :> t)
+      _             -> freshVar v
+    -- FIXME: This does not propagate the scope back!
```

Comment:
I think it's already correct. The scope due to the decls isn't relevant any more, now that they're wrapped up in a block. Rather than `runEmbedT`, you could just use `BuildScoped`, which scopes the decls and builds them into a block. (Also above.)

Reply:
Ohhh nice, I didn't know about `BuildScoped`! I think I agree that this is fine, as the code I generate here should have no free variables, and so no substitution should ever have any effect on its body (except for the subst from beta-reduction of course, but that's ok).